### ISSUE TITLE
where to start?

### ISSUE THREAD

**Issue by zakkirkharim** [CESNET/Netopeer2#149]
We are just starting to study about how we can use Netopeer or netopeer2 for supporting NETCONF interface to EMS and use the netopeer cli as cli.
Can you please point to documentation or page i can start from? Specifically, i would like to know whetehr the recommendation is to use netopeer2 if we are starting new. Also if we are not using netopeer2, what we will miss.
We had looked into the callback functions of libnetconf to write the code to made the application specific configuration. Will netopeer2 with sysrepo make this simpler?

**Comment by rkrejci** [comment]
Do not use libnetconf or Netopeer, they are obsolete. Instead, use [libnetconf2](https://github.com/CESNET/libnetconf2) and [Netopeer2](https://github.com/CESNET/Netopeer2). Maybe you would be interested also in [libyang](https://github.com/CESNET/libyang)

If you are going to integrate NETCONF support into your client application, I'd prefer to use directly libnetconf2 instead of netopeer2-cli. The CLI can serve to you as an example how to use libnetconf2. The documentation for libnetconf2 can be generated from source code, or it is available [online](https://netopeer.liberouter.org/doc/libnetconf2/) - start with the `How To ...` pages.

The callback mechanism was supposed for the server-side (device), so I'm not sure you will need it. Anyway, sysrepo already includes much better callback mechanism. So for the server-side integration start at [sysrepo.org](http://www.sysrepo.org/) and generated [docs](http://www.sysrepo.org/static/doc/html/)

**Comment by shehan492** [comment]
I have installed netopeer2-cli and netopeer2-server as found in the following instruction.
https://hub.docker.com/r/sysrepo/sysrepo-netopeer2/~/dockerfile/
cli and server is up and using the cli, I could connect the server. But most of the cases, when issuing get or get-config, the server crashes with the following log.

> netopeer2-server[3958]:
 All data nodes and constraints resolved.
netopeer2-server[3958]: Session 1: thread 1 event new RPC.
Segmentation fault, backtrace:
netopeer2-server(signal_handler+0x87)[0x406c2e]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x11390)[0x7f7e4cdff390]
netopeer2-server(op_sr_val_to_lyd_node+0x1c9)[0x4116b1]
netopeer2-server[0x4118e2]
netopeer2-server(op_get+0x629)[0x411f93]
/usr/local/lib/libnetconf2.so.0.9(+0x1f80a)[0x7f7e4c8b880a]
/usr/local/lib/libnetconf2.so.0.9(nc_ps_poll+0x60b)[0x7f7e4c8b956a]
netopeer2-server[0x409168]
/lib/x86_64-linux-gnu/libpthread.so.0(+0x76ba)[0x7f7e4cdf56ba]
/lib/x86_64-linux-gnu/libc.so.6(clone+0x6d)[0x7f7e4c1e33dd]
 
> 

No specific configurations were made while calling the get command. Attaching the server startup log and system details for reference.

Please suggest how to proceed.I'm currently blocked on this. My objective is to add custom yang modules based on my device configurations. 

[serverlog_And_system_details.txt](https://github.com/CESNET/Netopeer2/files/1135354/serverlog_And_system_details.txt)

**Comment by michalvasko** [comment]
Hi,
it would be best if you could run *netopeer2-server* with `valgrind` and then post the output, hard for me to know the problem based on this. Anyway, please create a new issue, there is no actual connection to this one and at least include what versions of all the applications you have used.

Regards,
Michal

**Comment by shehan492** [comment]
Hi Michal,

Now the crash is not observed when i reinstalled all packages with the correct versions of libyang (v0.12-r2), libnetconf2 (v0.8-r1) and sysrepo (v0.6.0) as mentioned in release info. When I add a sample yang module with sysrepo using command 

> sysrepoctl --install --yang=/home/user/ietf-interfaces.yang --owner=admin:admin  --permissions=664

 and do a get-config operation in cli, This is always coming as startup-config and not in running-config. I think I lack clarity in how to use sysrepo(Right now I'm going through the documentations)  and the netopeer-cli properly. Could you please direct me with some useful tutorials or the exact order of procedures to get this correct?
@michalvasko

**Comment by michalvasko** [comment]
Hi,
I guess it is best to read [this](http://www.sysrepo.org/static/doc/html/dev_guidelines.html) documentation section. Although it is lacking some information, you should learn what the problem is in your case.

Regards,
Michal

**Comment by shehan492** [comment]
Hi Michal,
Thanks for the reply. I went through the documents and was able to get the data store working properly. I have documented the complete steps that I followed for setting up sysrepo,netopeer2-cli and server. Attaching it here as it might be useful for someone. 

Regards,
Shahan. 
[netopeer2_procedure.pdf](https://github.com/CESNET/Netopeer2/files/1162083/netopeer2_procedure.pdf)

**Comment by shehan492** [comment]
Hi Michal,

I have another query. Please suggest if there is any java netconf client  that can be used along with netopeer2-server and sysrepo.
@michalvasko

**Comment by michalvasko** [comment]
Hi Shahan,
since I always work with our *netopeer2-cli*, I cannot recommend you any Java NETCONF client, you will have to find one yourself. Any that conforms to NETCONF RFCs should work with *netopeer2-server*.

Regards,
Michal
